File: lib/platform/apps/system_dashboard/static/projections/registry.js
// projections/registry.js
//
// The projection registry maps a "shape tag" (the result of `shapeOf(value)`)
// to a stack of projection functions. The active projection is the last one
// pushed via `register` or moved to the end via `select`. This lets future
// pack-shipped projections layer on top of built-ins without mutating them
// — a pack registers its own renderer for a given tag, and `select` makes it
// active. Unwinding (e.g. on pack unload) just splices it out.
//
// Ctx shape (consumed by projection functions; documented but not exported):
//   Ctx = {
//     project: (value) => Element | StreamConsumer,  // recurse via registry
//     action:  (alias_id, args?, onResult?) => clickHandler, // bound elsewhere
//   }

const reg = new Map(); // tag -> Projection[]

export function register(tag, fn) {
  if (typeof tag !== "string") throw new TypeError("register: tag must be string");
  if (typeof fn !== "function") throw new TypeError("register: fn must be function");
  let arr = reg.get(tag);
  if (!arr) { arr = []; reg.set(tag, arr); }
  arr.push(fn);
}

export function select(tag, fn) {
  // Move fn to the end of the array (active position) — caller must have
  // register()ed it first.
  const arr = reg.get(tag);
  if (!arr) throw new Error("select: no projections for tag " + tag);
  const idx = arr.indexOf(fn);
  if (idx === -1) throw new Error("select: fn not registered for tag " + tag);
  arr.splice(idx, 1);
  arr.push(fn);
}

export function has(tag) { return reg.has(tag); }

export function shapeOf(v) {
  if (v === null) return "null";
  if (Array.isArray(v)) return "array";
  const t = typeof v;
  if (t !== "object") return t;
  return typeof v.type === "string" ? v.type : "object";
}

export function project(v, ctx) {
  const tag = shapeOf(v);
  const arr = reg.get(tag);
  const fn = arr && arr.length > 0 ? arr[arr.length - 1] : null;
  if (!fn) {
    // Lazy default — `_default.js` registers itself via `setDefault`. Keeping
    // a plain function reference here breaks the circular import:
    //   registry.js -> _default.js -> registry.js
    return _defaultProjection(v, ctx);
  }
  return fn(v, ctx);
}

let _defaultProjection = (v) => {
  throw new Error("registry: no _default projection registered");
};

export function setDefault(fn) {
  if (typeof fn !== "function") throw new TypeError("setDefault: fn must be function");
  _defaultProjection = fn;
}

File: lib/platform/apps/system_dashboard/static/projections/registry.lua
-- lib/platform/apps/system_dashboard/static/projections/registry.lua
-- Type-stub for ./registry.js.
--
-- Initiative B Commit C. This file exists so projection Lua sources can write
--
--     local registry = require("lib.platform.apps.system_dashboard.static.projections.registry")
--     registry.register("text", project_text)
--
-- and the typechecker resolves the require through default `package.path`,
-- reads this file, and gets accurate types for every named export.
--
-- The implementations are no-ops that `error()` — these stubs are NEVER
-- executed at runtime. The transpile-time pipeline (lib/lua2ts) remaps the
-- require to `./registry.js` via `opts.imports` so the browser loads the
-- real registry directly from the JS module.
--
-- JS exports mirrored here:
--   register(tag, fn)     — push fn onto the stack for tag
--   select(tag, fn)       — move an already-registered fn to active position
--   has(tag) -> bool      — does tag have at least one projection?
--   shapeOf(v) -> string  — tag derivation: "null"/"array"/typeof/v.type
--   project(v, ctx)       — top-of-stack dispatch via shapeOf
--   setDefault(fn)        — install the fallback projection (used by _default.js)
--
-- See projection_types.lua for the canonical Ctx, Projection, and Element
-- type aliases. That file's declared globals (`dom`, `text`) stay in place
-- for now; Commit D removes them once projection sources migrate from the
-- declared-global style to require() everywhere.

--:: require "lib.web.js_types"
--:: require "lib.platform.apps.system_dashboard.primitive_types"
--:: require "lib.platform.apps.system_dashboard.projections.projection_types"

local M = {}

--: (tag: string, fn: Projection) -> ()
function M.register(_tag, _fn)
  error("registry.register: stub")
end

--: (tag: string, fn: Projection) -> ()
function M.select(_tag, _fn)
  error("registry.select: stub")
end

--: (tag: string) -> boolean
function M.has(_tag)
  error("registry.has: stub")
end

--: (v: unknown) -> string
function M.shapeOf(_v)
  error("registry.shapeOf: stub")
end

--: (v: Primitive, ctx: Ctx) -> Element
function M.project(_v, _ctx)
  error("registry.project: stub")
end

--: (fn: Projection) -> ()
function M.setDefault(_fn)
  error("registry.setDefault: stub")
end

return M
